### ISSUE TITLE
Possible False Positive in upcoming CRS, rule 932120

### ISSUE THREAD

**Issue by MirkoDziadzka** [coreruleset/coreruleset#3384]
### Description

Future CRS (nightly build) might contain a False Positive and will flag the word "powershell" as an attack in PL1
I think this should not be flagged by rule 932120

### How to reproduce the misbehavior (-> curl call)

Running the same request against the sandbox in 3.3.2 and nightly give different results for the string "powershell"

```
$ curl -H "x-format-output: txt-matched-rules" -H "x-crs-version: 3.3.2" https://sandbox.coreruleset.org/?s=powershell
```

and

```
$ curl -H "x-format-output: txt-matched-rules" -H "x-crs-version: nightly" https://sandbox.coreruleset.org/?s=powershell
932120 PL1 Remote Command Execution: Windows PowerShell Command Found
949110 PL1 Inbound Anomaly Score Exceeded (Total Score: 5)
980170 PL1 Anomaly Scores: (Inbound Scores: blocking=5, detection=5, per_pl=5-0-0-0, threshold=5) - (Outbound Scores: blocking=0, detection=0, per_pl=0-0-0-0, threshold=4) - (SQLI=0, XSS=0, RFI=0, LFI=0, RCE=5, PHPI=0, HTTP=0, SESS=0, COMBINED_SCORE=5)
```

### Logs

See sandbox

### Your Environment

See sandbox

### Confirmation

[x] I have removed any personal data (email addresses, IP addresses,
    passwords, domain names) from any logs posted.

**Comment by azurit** [comment]
@MirkoDziadzka Hi and thanks for reporting this. Why do you think this is a false positive?

**Comment by MirkoDziadzka** [comment]
@azurit I think the simple presence of the word `powershell` in an input should not be triggered in PL1. Same as "select" is not triggered as an SQLi. YMMV.

**Comment by dune73** [comment]
We may need to look into this. Thanks for sharing @MirkoDziadzka.

What is your take on this @theseion? Do we have options to add context in neighboring rules?

**Comment by theseion** [comment]
`powershell` was added as part of a bug bounty finding. If I'm not mistaken, we don't have any RCE rules above PL 1 for Windows. I understand that `powershell` makes much less sense at PL 1 then PowerShell commandlets, so I don't have anything against removing it. We would need to consider though, whether we wanted to add a new rule at Pl 2.

**Comment by fzipi** [comment]
Agreed, it is coming from this one: https://github.com/coreruleset/coreruleset/pull/2589. It was called `powershell.exe` before (see https://github.com/coreruleset/coreruleset/pull/2589/files#diff-bf277221b446105abad5fd3c89bdea6ac99b7b74269aad782669f969f8a535e7L1).

Per latest comment from Walter, let's remove it from PL1.

**Comment by dune73** [comment]
I agree on the removal from PL1, but I think we need to catch the naked keyword `powershell` at higher levels. If there is no rule where it fits, then I think we need a rule/place for this keyword and Windows friends at PL2 or PL3.

**Comment by MirkoDziadzka** [comment]
A genral question about 932120

To the best of my knowledge, you need some kind of separator before you can start a command injection. The pure command should not be the problem, at least some characters must come before. Not sure about windows, but on Unix I would assume that we need something like `'";&|)]` before the command to have an injection.

Is this different on Windows? Am I missing something?

**Comment by fzipi** [comment]
You are right. But this rule comes form a list using `pmFromFile`, where there is no regular expression involved. Therefore there is no previous chars that we can use, nor additional chars at the end.

**Comment by MirkoDziadzka** [comment]
@fzipi 

> You are right. But this rule comes form a list using `pmFromFile`, where there is no regular expression involved. Therefore there is no previous chars that we can use, nor additional chars at the end.

Then maybe we need the `@pmf`  + a chain rule with matches on any special character which can introduce an injection. Like `@rx []'";&|)}]` (No idea if this is complete) . This can still produce false positives, but better the current version. But I will leave it to the experts ;-) Thanks.

**Comment by dune73** [comment]
This is a good idea, but it's a slippery road that leads to evasions very quickly. What you have in the chained rule is the MATCHED_VAR, yet the first rule in the chain may run multiple times, resulting in a MATCHED_VAR that carries an additional naked `powershell` while the attack payload has been overwritten in the meantime. Very tricky business.

**Comment by EsadCetiner** [comment]
It looks like false positives for rule 932120 is pretty rare, I only ever encountered one false positive personally, `ftw quantitative` reports only reports three false positives out of a million requests, and reports from users are pretty rare too.
```
$ ftw quantitative -r 932120 --debug --max-concurrency 100 --corpus-size 1M
6:05PM INF ⏳Running quantitative tests with 100 goroutines
6:05PM DBG Using paranoia level: 4
6:05PM DBG Using default corpus cache directory: /home/esadc/.ftw
6:05PM DBG Preparing download of corpus file from https://downloads.wortschatz-leipzig.de/corpora/eng_news_2023_1M.tar.gz
6:05PM DBG filename eng_news_2023_1M-sentences.txt already exists
6:06PM DBG **> rule 932120 (PL1) with payload 167112 =>  #0: uri_payload:config maps and secrets that contain embedded powershell commands and are mounted into pods are also a strong indication of exploitation.”
6:08PM DBG **> rule 932120 (PL1) with payload 583719 =>  #0: uri_payload:powershell remoting is a feature that enables users to create and manage sessions on remote computers.
6:09PM DBG **> rule 932120 (PL1) with payload 816263 =>  #0: uri_payload:these include hackers abusing legitimate software running on the network such as powershell to exploiting zero day vulnerabilities something that was relatively rare until last year.
6:10PM DBG Printing Stats summary
Run 1000000 payloads (0 skipped) in 5m51.981872705s
Total False positive ratio: 3/1000000 = 0.0000
Total False positive sentences: 3/1000000 = 0.0000
False positives per paranoia level:
  PL1: 3 false positives. FP Ratio: 3/1000000 = 0.0000
False positives per rule id:
  932120 (PL1): 3 false positives. FP Ratio: 3/1000000 = 0.0000
```

There are a few false positive reports that were unresolved, but closed as can be seen here:
https://github.com/coreruleset/coreruleset/issues/1768
https://github.com/coreruleset/coreruleset/issues/3698
https://github.com/coreruleset/coreruleset/issues/517

All the examples are pretty niche and likely to be uncommon. Adding a word boundary via a chain rule should solve the most reported false positives for fairly little effort with little risk of false negatives, although the natural language false positives will be hard to remedy without significant effort (I don't think it's worth it fixing those considering it's so rare).

> This is a good idea, but it's a slippery road that leads to evasions very quickly. What you have in the chained rule is the MATCHED_VAR, yet the first rule in the chain may run multiple times, resulting in a MATCHED_VAR that carries an additional naked powershell while the attack payload has been overwritten in the meantime. Very tricky business.

@dune73 That shouldn't be an issue with [MATCHED_VARS](https://github.com/owasp-modsecurity/ModSecurity/wiki/Reference-Manual-(v2.x)#user-content-MATCHED_VARS) as it includes all the matched inputs, rather than the last one that was matched with `MATCHED_VAR`.

**Comment by HackingRepo** [comment]
yes powershell without any delimiter alone, is just a cmd arg, unless your app give someone execute arbiratry command in that case, you very stupid, if powershell not followed by a delimiter or args just a word and legitimate
